packages/babel-plugin-transform-for-of/src/index.js

```diff
@@ -189,156 +160,74 @@ export default declare((api, options) => {
           right.isGenericType("Array") ||
           t.isArrayTypeAnnotation(right.getTypeAnnotation())
         ) {
-          replaceWithArray(path);
+          path.replaceWith(_ForOfStatementArray(path));
           return;
         }
 
-        const { node } = path;
-        const build = pushComputedProps(path, state);
-        const declar = build.declar;
-        const loop = build.loop;
-        const block = loop.body;
+        if (!state.availableHelper(builder.helper)) {
+          // Babel <7.9.0 doesn't support this helper
+          transformWithoutHelper(loose, path, state);
+          return;
+        }
+
+        const { node, parent, scope } = path;
+        const left = node.left;
+        let declar;
+
+        const stepKey = scope.generateUid("step");
+        const stepValue = t.memberExpression(
+          t.identifier(stepKey),
+          t.identifier("value"),
+        );
+
+        if (
+          t.isIdentifier(left) ||
```

Comment:
nit: It suffices to check `t.isVariableDeclaration` first and assuming it's `LVal` in the else branch, because we have validated `ForOfStatement.left` even when `BABEL_TYPES_8_BREAKING` is not enabled:

https://github.com/babel/babel/blob/1ba41f208430bdf382b75b56e9aad7436ca522ad/packages/babel-types/src/definitions/es2015.js#L333